### DISCUSSION TITLE
Best practice for redirect URL for native applications

### DISCUSSION THREAD

**Discussion by derberni** [ory/hydra#2368] [Q&A] [answered]
Hi,
we plan on using Hydra with a native application as a client and are wondering how to best handle the redirect to the client that contains the access code. According to the best practices [https://www.rfc-editor.org/rfc/rfc8252.txt](RFC8252) there are three possibilities, but we are unsure which of these are supported by Hydra.

It seems that Hydra only allows https redirect URLs, unless the --dangerous-allow-insecure-redirect-urls strings command line flag is provided. So without that flag, redirecting to localhost is not an option. This would have actually been our preferred choice, because our application ships for several different operating systems, and the other proposed method of using a custom uri scheme will be hard to implement on all operating systems.

Thanks in advance,
David

**Comment by vinckr** [comment, marked as answer]
All of these are supported in Ory Hydra, but as you pointed out we definitely recommend using https redirect. 
You can use `--dangerous-allow-insecure-redirect-urls` for development, but do not use it in production! 

I understand that this makes your problem harder, but a central "mantra" of the Ory ecosystem is to make insecure things hard.
Is redirecting to an IP an option instead of localhost?

**Reply by derberni** [reply]
Thx for the quick reply. The "mantra" is definitely a more secure approach, so I can totally understand it ;) I don't yet grasp how an IP instead of localhost could help. Could you please share more details of this approach?

We also thought about using an additional service that is inspired by the device grant and this article: [Add the OAuth 2.0 Device Flow to any OAuth Server](https://developer.okta.com/blog/2019/02/19/add-oauth-device-flow-to-any-server). The additional service would be the registered oauth2 client with a valid https redirect url. Our application would be able to ask the service for an authorization URL and redirect the user to that URL. After completing login/consenst flow, the access code would be sent to the service, which would cache it until our application queries it from the service (for example by polling like in the device grant).
My main concern with this approach is, that anybody would be able to figure out the service endpoints for getting the authorization URL and for polling for the access_code, so they would be able to easily impersonate our application. But I guess that is a general problem with native applications, since you will always be able to get a client id and/or secret out of the compiled code.

**Comment by vinckr** [comment]
My bad, nevermind the localhost/IP, that wont work without the flag.

Do you mean this "Device Authorization Grant"? (https://tools.ietf.org/html/rfc8628)
Related issue: https://github.com/ory/hydra/issues/1553.

I am not sure if I understand the second approach.
Is your use case possible at all when using a custom uri scheme? 
If not maybe you could open an issue, describing your use case in detail and with some info on why it is not working at the moment.

ps: I found this issue for the Loopback interface redirection : https://github.com/ory/fosite/pull/400

**Reply by Nilaik** [reply]
Sorry for the late reply. Actually it was our bad, we somehow missed that for localhost and the IP 127.0.0.1 it is not necessary to set the flag `--dangerous-allow-insecure-redirect-urls`. So it is simply possible to use 127.0.0.1 callback addresses with http also in a production environment. This is what we will be using for our application.

The 'Device Authorization Grant' you have linked is what we were looking at, too. Thanks for sharing the related issue. For the moment we won't need it but it might get interesting again for us later on. Especially for remote access of our application where we need to login and the callback to localhost will not work since we are on a different device.

Again thanks for your help!

**Reply by vinckr** [reply]
Ok great 🙂 
If you are open to it, I would love to hear some more about your use case, feel free to contact me via email/Slack !
There also has been some more input on it from interested parties in the community: https://github.com/ory/hydra/issues/2416 , so that could be relevant for you as well.
